configcatclient/lazyloadingcachepolicy.py

```diff
@@ -43,7 +51,6 @@ def get(self):
 
     def force_refresh(self):
         try:
-            self._lock.acquire_write()
```

Comment:
I would keep this lock here. I would make a "private" _force_refresh() function without lock and call that function here with the acquire_write and release_write.
I'll add a separate comment with the whole code.

Reply:
```
class LazyLoadingCachePolicy(CachePolicy):

    def __init__(self, config_fetcher, config_cache, cache_time_to_live_seconds=60):
        if cache_time_to_live_seconds < 1:
            cache_time_to_live_seconds = 1
        self._config_fetcher = config_fetcher
        self._config_cache = config_cache
        self._cache_time_to_live = datetime.timedelta(seconds=cache_time_to_live_seconds)
        self._lock = ReadWriteLock()
        self._last_updated = None

    def get(self):
        try:
            self._lock.acquire_read()

            utc_now = datetime.datetime.utcnow()
            if self._last_updated is not None and self._last_updated + self._cache_time_to_live > utc_now:
                config = self._config_cache.get()
                if config is not None:
                    return config
        finally:
            self._lock.release_read()

        try:
            self._lock.acquire_write()
            # If while waiting to acquire the write lock another
            # thread has updated the content, then don't bother requesting
            # to the server to minimise time.
            if self._last_updated is None or self._last_updated + self._cache_time_to_live <= datetime.datetime.utcnow():
                self._force_refresh()
        finally:
            self._lock.release_write()

        try:
            self._lock.acquire_read()
            config = self._config_cache.get()
            return config
        finally:
            self._lock.release_read()

    def force_refresh(self):
        try:
            self._lock.acquire_write()
            self._force_refresh()
        finally:
            self._lock.release_write()

    def stop(self):
        pass

    def _force_refresh(self):
        try:
            configuration_response = self._config_fetcher.get_configuration_json()
            # set _last_updated regardless of whether the cache is updated
            # or whether a 304 not modified has been sent back as the content
            # we have hasn't been updated on the server so not need
            # for subsequent requests to retry this within the cache time to live
            self._last_updated = datetime.datetime.utcnow()
            if configuration_response.is_fetched():
                configuration = configuration_response.json()
                self._config_cache.set(configuration)
        except HTTPError as e:
            log.error('Double-check your SDK Key at https://app.configcat.com/sdkkey.'
                      ' Received unexpected response: %s' % str(e.response))
        except:
            log.exception(sys.exc_info()[0])
```

Reply:
@kp-cat I am not sure how well that works, since you would be getting the write lock twice and would never release it. I may be missing something here.

Reply:
@tyroneabdy I think it will work because the get function calls the inner/private _force_refresh() function. In the _force_refresh() function, there is no lock so when the get function calls it we can surround it with acquire_write()/acquire_release() without calling lock twice.

Besides that, if we call the "public" force_refresh() function it will call the inner/private _force_refresh() with acquire_write()/acquire_release() too.

Reply:
@kp-cat right yeah I see now, I missed that, that it was calling the private one in get() and that you could still call force_refresh() the public method.